### DISCUSSION TITLE
Using prior distributions

### DISCUSSION THREAD

**Discussion by kurokida** [kurokida/jsQuestPlus#5] [Ideas]
Hi, @tpronk 

First, let me explain an important change in the initialization of jsQuestPlus.

# About change in the initialization

## Before change

```javascript
    const jsqp = new jsQuestPlus(
        [func_resp0, func_resp1], [stim_domain], [threshold_domain, slope_domain, guess_domain, lapse_domain])
```

## After change

```javascript
    const jsqp = new jsQuestPlus({
        psych_func: [func_resp0, func_resp1],
        stim_samples: [stim_domain],
        psych_samples: [threshold_domain, slope_domain, guess_domain, lapse_domain],
    })
```

These changes are necessary in order to specify the prior distribution. Sorry, but we also need to make changes to our paper.

# Prior distributions

In previous jsQuestPlus, the prior for the psychometric parameters was uniformly distributed. Basically, that is fine, but the experimenter may want to specify the prior based on previous research or preliminary observations. I've updated jsQuestPlus for these requests.

## Specify the prior distribution individually

See https://github.com/kurokida/jsQuestPlus/blob/dev/demos/prior1.html

```javascript
// Example calculation for the prior PDF of threshold.
const tmp = jsQuestPlus.linspace(1, threshold_samples.length)
const threshold_prior = jsQuestPlus.div(tmp, jsQuestPlus.sum(tmp))

const jsqp = new jsQuestPlus({
    psych_func: [func_resp0, func_resp1],
    stim_samples: [contrast_samples],
    psych_samples: [threshold_samples, slope_samples, guess, lapse_samples],
    priors: [threshold_prior, 1, 1, 1] // This should match the order of the psych_samples. To specify a uniform distribution (default), write 1.
})
```

## Use the previous jsQuestPlus data as prior distribution

See https://github.com/kurokida/jsQuestPlus/blob/dev/demos/prior2.html

```javascript
const prior = jsqp1.normalized_posteriors // Store the PDF of condition 1 for condtion 2

const jsqp2 = new jsQuestPlus({
    psych_func: [func_resp0, func_resp1],
    stim_samples: [contrast_samples],
    psych_samples: [threshold_samples, slope_samples, guess, lapse_samples],
    mult_prior: prior // Specify a prior probability distribution
})
```

If you run https://github.com/kurokida/jsQuestPlus/blob/dev/demos/prior3.html, you'll see the estimated values of the parameters become more and more accurate with each simulation when using the previous jsQuestPlus data as prior distribution.

Sorry for the brief explanation.
If you have any questions, please don't hesitate to let me know.

**Comment by tpronk** [comment]
I love the named parameters; that makes the configuration much clearer!

One should specify either `priors` or `mult_prior` right? To prevent any confusion between these two, I've got two ideas (which are mutually exclusive)

**Option 1.** Add an error message if both are specificied?

**Option 2.** Have a only the `mult_prior` argument, but add a function for normalizing the priors? Something like...

Specify distributions individually
```
const jsqp = new jsQuestPlus({
    psych_func: [func_resp0, func_resp1],
    stim_samples: [contrast_samples],
    psych_samples: [threshold_samples, slope_samples, guess, lapse_samples],
    mult_priors: jsquest.normalize([threshold_prior, 1, 1, 1])
})
```

Use posterior distribution of a previous run of jsQuestPlus
```
const jsqp = new jsQuestPlus({
    psych_func: [func_resp0, func_resp1],
    stim_samples: [contrast_samples],
    psych_samples: [threshold_samples, slope_samples, guess, lapse_samples],
    mult_priors: prior
})
```

**Reply by kurokida** [reply]
Thanks for the great idea.
I would really like to adopt your method, but I don't know how to change the code.

A uniform value (1 when summed) is set from [line 46](https://github.com/kurokida/jsQuestPlus/blob/4da3693d9d4c7ccd052176d0f9f2419eba8e6fb3/src/jsQuestPlus.js#L46
) to [line 55](https://github.com/kurokida/jsQuestPlus/blob/4da3693d9d4c7ccd052176d0f9f2419eba8e6fb3/src/jsQuestPlus.js#L55) as the initial value for the prior distribution.

When priors are specified, the number of elements in the array is check for the same as the number of elements in the array of the psychometric parameter in [line 60](https://github.com/kurokida/jsQuestPlus/blob/4da3693d9d4c7ccd052176d0f9f2419eba8e6fb3/src/jsQuestPlus.js#L60), and the prior distribution is overwritten in [line 62](https://github.com/kurokida/jsQuestPlus/blob/4da3693d9d4c7ccd052176d0f9f2419eba8e6fb3/src/jsQuestPlus.js#L62).

Finally, *priors* should be assigned to *this.priors* in [line 67](https://github.com/kurokida/jsQuestPlus/blob/4da3693d9d4c7ccd052176d0f9f2419eba8e6fb3/src/jsQuestPlus.js#L67).

In other words, it would be nice if I could create a new normalize function and access the *this* parameter in that function, but I don't know how to do that.

Could you please commit the code?

**Reply by tpronk** [reply]
Sure I'd happily do some coding. Before I go ahead, let me check something: The normalization part doesn't need to know how many elements should be in priors, just scale them so that they add up to 1. Hence, `normalize` can be a static method. Then the check whether the number of priors is equal to can be performed in the jsQuestPlus constructor.

Did I get this right?

**Reply by kurokida** [reply]
Thank you for your cooperation!
It is difficult for me to explain in English, but I will try.

First, regarding your suggestion of *jsquest.normalize*, the name may not be appropriate

Let's take an example where the length of parameter 1 is 4, the length of parameter 2 is 3, and the length of parameter 3 is 2.
In this case, the default prior distribution of each parameter is as follows
parameter1 -> [0.25 0.25 0.25 0.25]
parameter2 -> [0.33 0.33 0.33]
paramter3 -> [0.5 0.5]

The mult_priors will be a combination of all of these values, multiplied by three parameters (e.g., 0.25 x 0.33 x 0.5).
The length of the mult_priors is 4 x 3 x 2=24
This value will be assigned to mulitiplied_priors, which is initialized in [line 70](https://github.com/kurokida/jsQuestPlus/blob/4da3693d9d4c7ccd052176d0f9f2419eba8e6fb3/src/jsQuestPlus.js#L70).
The mulitiplied_priors will be normalized in [line 87](https://github.com/kurokida/jsQuestPlus/blob/4da3693d9d4c7ccd052176d0f9f2419eba8e6fb3/src/jsQuestPlus.js#L87); the *jsquest.normalize* should return this value.

If the experimenter specify the prior for a certain psychometric parameter, the function should check whether the length of the prior is same as that of the parameter samples.
In addition, all combinations of prior distributions should be stored to the jsQuestPlus data, as shown in [line 68](https://github.com/kurokida/jsQuestPlus/blob/4da3693d9d4c7ccd052176d0f9f2419eba8e6fb3/src/jsQuestPlus.js#L68).
So, I don't think this function can be made static.

Perhaps I haven't explained it properly. Please feel free to ask questions.

**Reply by kurokida** [reply]
[The explanation of the combvec in Matlab](https://jp.mathworks.com/help/deeplearning/ref/combvec.html?lang=en) may be useful.
I used the jsquest.combvec in [line 68](https://github.com/kurokida/jsQuestPlus/blob/4da3693d9d4c7ccd052176d0f9f2419eba8e6fb3/src/jsQuestPlus.js#L68)

**Reply by kurokida** [reply]
After I explained it to you, I came up with an idea. If you haven't started creating the code yet, I'll do it.

**Reply by tpronk** [reply]
I haven't yet, so cool; do your magic :)

**Comment by kurokida** [comment]
Hi, @tpronk 

I've finished updating the code.

Now, the *set_prior* function works.

## How to specify prior distributions individually

```javascript
    // Example calculation for the prior PDF of threshold.
    // Assume a Gaussian function with a mean of -18 and a standard deviation of 2.
    const threshold_prior = jsQuestPlus.gauss(threshold_samples, -18, 2)

    const jsqp = new jsQuestPlus({
        psych_func: [func_resp0, func_resp1],
        stim_samples: [contrast_samples],
        psych_samples: [threshold_samples, slope_samples, guess, lapse_samples],
        priors: jsQuestPlus.set_prior([threshold_prior, slope_samples.length, guess.length, lapse_samples.length]),
        // The order of the priors must match the order of the psych_samples. To specify a uniform distribution (default), write the length of the parameter.
    })

```
## How to use the previous jsQuestPlus data as a prior distribution

In this case, the set_prior function is not needed.

```javascript
    const jsqp2 = new jsQuestPlus({
        psych_func: [func_resp0, func_resp1],
        stim_samples: [contrast_samples],
        psych_samples: [threshold_samples, slope_samples, guess, lapse_samples],
        priors: jsqp1.posteriors // Specify the posteriors of the previous condition.
    })
```

I would like to confirm, regarding the following English name, is it appropriate?

## Property names

- psych_func
- stim_samples
- psych_samples
- priors

## Function names

- set_prior

## File names

- demos/specify_prior_individually.html
- demos/specify_the last_result_as_the_prior.html
- demos/Simlate_the_effect_of_prior.html
